control.lua

```diff
@@ -1,179 +1,496 @@
+      entity = entity,
+      is_working = is_working_function,
+      sec_avg = { capacity = 60, values = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}, next_index = 1, total = 0},
+      min_avg = { capacity = 60, values = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}, next_index = 1, total = 0},
+    }
     if global.show_labels then
-      local percent = math.max(math.floor(global.min_avg[id]:avg()*100), 0)
-      entity.surface.create_entity{name = "statictext", position = label_position_for(entity), text = tostring(percent).."%"}
+      add_label(data)
     end
+    global.entity_data[id] = data
+    return true
+  end
+  return false
+end
+
+--- Post-Processes the movement of the given entity.
+--
+-- @param entity:Entity - The entity that was moved.
+--
+local function move_entity(entity)
+  local data = global.entity_data[entity.unit_number]
+  if data and data.label and data.label.valid then
+    local position = label_position_for(entity)
+    data.label.teleport(position)
   end
 end
 
-function add_entity(entity)
-  local id = entity.unit_number
-  global.entities[id] = entity
-  global.entity_types[id] = entity.type
-  global.entity_positions[id] = entity.position
-  global.sec_avg[id] = MovingAvg(60)
-  global.min_avg[id] = MovingAvg(60)
+--- Remove the given entity from being tracked by UM.
+--
+-- @param id:uint - The id of the entity to stop tracking.
+--
+local function remove_entity(id)
+  local data = global.entity_data[id]
+  global.entity_data[id] = nil -- Prevent memory leaks
+  -- Remove label
+  remove_label(data)
 end
 
-function reset()
+--- Hard resets all data used by UM.
+--
+local function reset()
   game.print("UtilizationMonitor: Full reset")
-  global.version = VERSION
-  global.entities = {}
-  global.entity_types = {}
-  global.entity_positions = {}
+  -- clean up data used in earlier mod versions to make savegames smaller
+  global.entities = nil
+  global.entity_types = nil
+  global.entity_positions = nil
+  global.sec_avg = nil
+  global.min_avg = nil
+  global.last_mining_progress = nil
+  global.last_lab_durability = nil
+  -- end cleanup
 
-  global.sec_avg = {}
-  global.min_avg = {}
-  global.last_mining_progress = {}
-  global.last_lab_durability = {}
-  global.show_labels = settings.global["utilization-monitor-enabled"].value
+  global.version = VERSION
+  global.last_id = nil
+  -- Purge old labels
+  if global.entity_data ~= nil then
+    for _, data in pairs(global.entity_data) do
+      remove_label(data)
+    end
+  end
+  global.entity_data = {}
+  global.iteration = 0
+  global.show_labels = settings.global["utilization-monitor-show-labels"].value
+  global.entities_per_tick = settings.global["utilization-monitor-entities-per-tick"].value
+  global.iterations_per_update = settings.global["utilization-monitor-entities-per-tick"].value
```

Comment:
Isn't this supposed to be `utilization-monitor-iterations-per-update` here?

Reply:
Fixed